src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -78,37 +77,101 @@ AppHost::~AppHost()
+        const auto message = _logic.EarlyExitMessage();
+        if (!message.empty())
+        {
+            const auto displayHelp = result == 0;
+            const auto messageTitle = displayHelp ? IDS_HELP_DIALOG_TITLE : IDS_ERROR_DIALOG_TITLE;
+            const auto messageIcon = displayHelp ? MB_ICONWARNING : MB_ICONERROR;
+            // TODO:GH#4134: polish this dialog more, to make the text more
+            // like msiexec /?
+            MessageBoxW(nullptr,
+                        message.data(),
+                        GetStringResource(messageTitle).data(),
+                        MB_OK | messageIcon);
+
+            ExitProcess(result);
+        }
+    }
+}
+
+// Method Description:
+// - Retrieve the command line.
+// - Use our own algorithm to tokenize it because `CommandLineToArgvW` treats \"
+//   as an escape sequence to preserve the quotation mark (GH#4571)
+// - Populate the arguments as vector of hstrings.
+// Arguments:
+// - <none>
+// Return Value:
+// - Program arguments as vector of hstrings. If the function fails it returns an empty vector.
+std::vector<winrt::hstring> AppHost::_GetArgs() const noexcept
+{
+    try
+    {
+        std::vector<winrt::hstring> args{};
+        // get the command line
+        const wchar_t* const cmdLnPtr{ GetCommandLineW() };
+        if (!cmdLnPtr)
+        {
+            return {};
+        }
 
-        // Get the argv, and turn them into a hstring array to pass to the app.
-        wil::unique_any<LPWSTR*, decltype(&::LocalFree), ::LocalFree> argv{ CommandLineToArgvW(commandline, &argc) };
-        if (argv)
+        std::wstring cmdLn{ cmdLnPtr };
+        auto cmdLnEnd{ cmdLn.end() }; // end of still valid content in the command line
+        auto iter{ cmdLn.begin() }; // iterator pointing to the current position in the command line
+        auto argBegin{ iter }; // iterator pointing to the begin of an argument
+        auto quoted{ false }; // indicates whether a substring is quoted
+        auto within{ false }; // indicates whether the current character is inside of an argument
+
+        args.reserve(gsl::narrow_cast<size_t>(64U)); // pre-allocate some space for the string handles
+        while (iter < cmdLnEnd)
         {
-            std::vector<winrt::hstring> args;
-            for (auto& elem : wil::make_range(argv.get(), argc))
+            auto increment{ true }; // used to avoid iterator incrementation if a quotation mark has been removed
+            switch (*iter)
             {
-                args.emplace_back(elem);
+            case L' ': // space and tab are the usual separators for arguments
+            case L'\t':
+                if (!quoted && within)
+                {
+                    within = false;
+                    args.emplace_back(std::wstring{ argBegin, iter });
+                }
+                break;
+
+            case L'"': // quotation marks need special handling

```

Comment:
do we need to support commandlines that contain `\"`? what about `'`? honest question: i don't know

Reply:
> do we need to support commandlines that contain `\"`?

Well, the `\"` makes CommandlineToArgvW to remove the `\` and keeps the `"`.
E.g the result of `wt -d "c:\"` is argv[0] `wt`, argv[1] `-d`, and argv[2] `c:"` which leads to fail.
In any other case the `"`s are already removed entirely. Even weird command lines like
`wt -d c:\program" "files` will work because of that.
So the question is rather if we need an alternative way to preserve a quotation mark from getting removed? Such like doubling it as in VBS strings. I already raised my hand in #4571 because of that. Based on the [spec](https://github.com/microsoft/terminal/blob/master/doc/specs/%23607%20-%20Commandline%20Arguments%20for%20the%20Windows%20Terminal.md) I didn't implement anything like that in the parser since the answer seems to be no, we don't. @zadjii-msft ?

> what about `'`? 

Dunno. This is not affected by the backslash.
![image](https://user-images.githubusercontent.com/46659645/75488309-60b7dd00-59b0-11ea-855c-f2af3903e37b.png)

Reply:
Sorry, quick note while I'm between meetings. the `'` there is handled by powershell, isn't it? Maybe I don't totally understand.

What if you do `wt --% 'C:\program files\` ?

Reply:
> What if you do `wt --% -d 'C:\program files\` ?

It fails as expected. So does `wt --% -d 'C:\program files` and `wtd --% -d 'C:\program files\`.
I don't think that I changed anything but `\"` resulting in `\` rather than `"`. Certainly the `'` is handled by PowerShell in these examples. However, it is handled the same as before ¯\_(ツ)_/¯¯
I created folder `a'b` in the C: root and called the terminal app from within PowerShell
```
wt -d "c:\a'b"
wt -d c:\a`'b
```
Both lines are working with either `wt` or `wtd` (the latter with the own parser).

Reply:
@DHowett-MSFT If you want any special behavior for single quotes rather than the default then just tell me. Right now I'm a little confused if I even got your point 🤔

Reply:
@german-one sorry, I thought a lot about my question and realized I have no idea what I'm asking you for. I guess that explains your confusion, too! :smile: